File: server/internal/protocol/orchestration_events.go
package protocol

import "encoding/json"

// EventBase holds fields common to all domain events.
type EventBase struct {
	Sequence         int           `json:"sequence"`
	EventId          EventId       `json:"eventId"`
	Type             string        `json:"type"`
	AggregateKind    string        `json:"aggregateKind"`
	AggregateId      string        `json:"aggregateId"`
	OccurredAt       string        `json:"occurredAt"`
	CommandId        *CommandId    `json:"commandId,omitempty"`
	CausationEventId *EventId      `json:"causationEventId,omitempty"`
	CorrelationId    *CommandId    `json:"correlationId,omitempty"`
	Metadata         EventMetadata `json:"metadata"`
}

// EventMetadata carries optional provider-level tracing information.
type EventMetadata struct {
	ProviderTurnId *string            `json:"providerTurnId,omitempty"`
	ProviderItemId *ProviderItemId    `json:"providerItemId,omitempty"`
	AdapterKey     *string            `json:"adapterKey,omitempty"`
	RequestId      *ApprovalRequestId `json:"requestId,omitempty"`
	IngestedAt     *string            `json:"ingestedAt,omitempty"`
}

// OrchestrationEvent is a domain event with its payload kept as raw JSON.
// The bridge passes events through without deserializing every payload.
type OrchestrationEvent struct {
	EventBase
	Payload json.RawMessage `json:"payload"`
}

// ThreadMessageSentPayload is the payload for thread message-sent events.
// The bridge inspects this to relay message content.
type ThreadMessageSentPayload struct {
	ThreadId  ThreadId  `json:"threadId"`
	MessageId MessageId `json:"messageId"`
	Role      string    `json:"role"`
	Text      string    `json:"text"`
	TurnId    TurnId    `json:"turnId"`
	Streaming bool      `json:"streaming"`
}

// ThreadSessionSetPayload is the payload for thread session-set events.
// The bridge inspects this to track session state.
type ThreadSessionSetPayload struct {
	ThreadId     ThreadId `json:"threadId"`
	Status       string   `json:"status"`
	ProviderName string   `json:"providerName"`
	RuntimeMode  string   `json:"runtimeMode"`
	ActiveTurnId TurnId   `json:"activeTurnId"`
	LastError    string   `json:"lastError"`
}

File: mobile/composeApp/src/commonMain/kotlin/io/j3mobile/protocol/OrchestrationEvents.kt
package io.j3mobile.protocol

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonElement

/** Optional provider-level tracing information on events. */
@Serializable
data class EventMetadata(
    val providerTurnId: String? = null,
    val providerItemId: ProviderItemId? = null,
    val adapterKey: String? = null,
    val requestId: ApprovalRequestId? = null,
    val ingestedAt: String? = null,
)

/** A domain event with its payload kept as raw JSON. */
@Serializable
data class OrchestrationEvent(
    val sequence: Int,
    val eventId: EventId,
    val type: String,
    val aggregateKind: String,
    val aggregateId: String,
    val occurredAt: String,
    val commandId: CommandId? = null,
    val causationEventId: EventId? = null,
    val correlationId: CommandId? = null,
    val metadata: EventMetadata = EventMetadata(),
    val payload: JsonElement? = null,
)

/** Payload for thread message-sent events. */
@Serializable
data class ThreadMessageSentPayload(
    val threadId: ThreadId,
    val messageId: MessageId,
    val role: String,
    val text: String,
    val turnId: TurnId,
    val streaming: Boolean,
)

/** Payload for thread session-set events. */
@Serializable
data class ThreadSessionSetPayload(
    val threadId: ThreadId,
    val status: String,
    val providerName: String,
    val runtimeMode: String,
    val activeTurnId: TurnId,
    val lastError: String,
)

/** Payload for thread activity-appended events. */
@Serializable
data class ThreadActivityAppendedPayload(
    val threadId: ThreadId,
    val activityId: EventId,
    val tone: String,
    val kind: String,
    val summary: String,
    val payload: JsonElement? = null,
    val turnId: TurnId? = null,
)
